kani-compiler/src/codegen_cprover_gotoc/codegen/function.rs

```diff
@@ -345,64 +254,27 @@ impl<'tcx> GotocCtx<'tcx> {
         let def_id = self.current_fn().instance().def_id();
-        if self.is_test_harness_closure(def_id) {
-            let loc = self.codegen_span(&self.current_fn().mir().span);
-            self.test_harnesses.push(HarnessMetadata {
-                pretty_name: self.current_fn().readable_name().to_owned(),
-                mangled_name: self.current_fn().name(),
-                crate_name: self.current_fn().krate(),
-                original_file: loc.filename().unwrap(),
-                original_start_line: loc.start_line().unwrap() as usize,
-                original_end_line: loc.end_line().unwrap() as usize,
-                solver: None,
-                unwind_value: None,
-                // We record the actual path after codegen before we dump the metadata into a file.
-                goto_file: None,
-            })
+        if is_test_harness_closure(self.tcx, def_id) {
+            self.test_harnesses.push(self.default_kanitool_proof())
         }
     }
 
     /// This updates the goto context with any information that should be accumulated from a function's
     /// attributes.
     ///
     /// Handle all attributes i.e. `#[kani::x]` (which kani_macros translates to `#[kanitool::x]` for us to handle here)
-    fn handle_kanitool_attributes(&mut self) {
+    fn record_kani_attributes(&mut self) {
         let def_id = self.current_fn().instance().def_id();
-        let all_attributes = self.tcx.get_attrs_unchecked(def_id);
-        let (proof_attributes, other_attributes) = partition_kanitool_attributes(all_attributes);
-        if !proof_attributes.is_empty() {
-            self.create_proof_harness(other_attributes);
-        }
-    }
-
-    /// Create the proof harness struct using the handler methods for various attributes
-    fn create_proof_harness(&mut self, other_attributes: Vec<(String, &Attribute)>) {
-        let mut harness = self.default_kanitool_proof();
-        for attr in other_attributes.iter() {
-            match attr.0.as_str() {
-                "solver" => self.handle_kanitool_solver(attr.1, &mut harness),
-                "stub" => {
-                    if !self.queries.get_stubbing_enabled() {
-                        self.tcx.sess.span_warn(
-                            attr.1.span,
-                            "Stubbing is not enabled; attribute `kani::stub` will be ignored",
-                        )
-                    }
-                }
-                "unwind" => self.handle_kanitool_unwind(attr.1, &mut harness),
-                _ => {
-                    self.tcx.sess.span_err(
-                        attr.1.span,
-                        format!("Unsupported Annotation -> {}", attr.0.as_str()).as_str(),
-                    );
-                }
-            }
+        let attributes = extract_harness_attributes(self.tcx, def_id);
+        if let Some(attrs) = attributes {
+            let mut harness = self.default_kanitool_proof();
```

Comment:
Nit: It would be slightly cleaner to add another method that takes attributes as a parameter and use it to set the attributes field.

Reply:
Totally! What if I replace `default_kanitool_proof(&self)` by something like `generate_metadata(&self, attributes: Option<HarnessAttributes>` and set the attributes as I'm creating the metadata?

Reply:
Sounds good.